networkx/algorithms/dag.py

```
@@ -162,89 +162,166 @@ def topological_sort(G, nbunch=None, reverse=False):
 
     Raises
     ------
     NetworkXError
-        Topological sort is defined for directed graphs only. If the
-        graph G is undirected, a NetworkXError is raised.
+        Topological sort is defined for directed graphs only. If the graph G
+        is undirected, a NetworkXError is raised.
 
     NetworkXUnfeasible
-        If G is not a directed acyclic graph (DAG) no topological sort
-        exists and a NetworkXUnfeasible exception is raised.
+        If G is not a directed acyclic graph (DAG) no topological sort exists
+        and a NetworkXUnfeasible exception is raised.
 
     Notes
     -----
-    This is a recursive version of topological sort.
+    This algorithm is based on a description and proof in
+    Introduction to algorithms - a creative approach [1]_ .
 
     See also
     --------
     topological_sort
-    is_directed_acyclic_graph
 
+    References
+    ----------
+    .. [1] Manber, U. (1989). Introduction to algorithms - a creative approach. Addison-Wesley.
+        http://www.amazon.com/Introduction-Algorithms-A-Creative-Approach/dp/0201120372
     """
     if not G.is_directed():
         raise nx.NetworkXError(
             "Topological sort not defined on undirected graphs.")
 
-    def _dfs(v):
-        ancestors.add(v)
+    if key is None:
+        key = lambda x: x
 
-        for w in G[v]:
-            if w in ancestors:
-                raise nx.NetworkXUnfeasible("Graph contains a cycle.")
+    # nonrecursive version
+    indegree_map = {node: len(G.pred[node])
+                    for node in G}
+    # This integer breaks ties since node names only have to be hashable
+    # -- not comparable.  Python 3.0 introduces the nonlocal keyword;
+    # Python 2.7 forces us to work around it using a map.
+    tiebreaker = {None: 0}
 
-            if w not in explored:
-                _dfs(w)
+    def create_tuple(node):
+        retval = key(node), tiebreaker[None], node
```

Actually I think it is better to raise that error.  It's a corner case but good to know if you hit it.